### PR TITLE
DOC: Numpy docstrings migration

### PR DESCRIPTION
**Pull request by VeckoTheGecko**
This PR focusses on the following features (in order of implementation):
- [x] Initial conversion of `rst` docstrings to `numpydoc` format (with the help of pyment)
- [x] `numpydoc` docstring validators to ensure docstring quality
  - [x] Install `numpydoc` docstring validators as part of the Sphinx build process
  - [x] Ensure docstrings are compliant with chosen validators
  - [x] Tying in docstring validation with CI/CD
- [x] Add types to docstring parameters (particularly focussing on popular modules).

PR contributes to #1324, where you can read more on this featureset.

Draft PR for visibility (and to trigger RTD builds).

(given the size of these diffs, let me know if you want me to break into multiple PRs)

### CODE REVIEW

**Comment by VeckoTheGecko** [comment]
Done with initial conversion of docstrings to numpy format. Now moving onto docstring validation.

**Comment by VeckoTheGecko** [comment]
Numpydoc validators have been included in the build process (basically linting through docstrings, looking for things like "is the argument in the docstring the same as the one in the function call", ensuring the docstring is rendered correctly).

The full list of possible validation checks are outlined in the [error codes](https://numpydoc.readthedocs.io/en/latest/validation.html). Some of the codes are incompatible with the codebase (either outright, or due to the state of many docstrings). Implementing all these codes is a very very low priority, but the small subset of codes at the moment are a good starting point (and additional codes can be included easily down the line if desired).

**Comment by VeckoTheGecko** [comment]
Just confirming @erikvansebille , is this option checked in the `Advanced Settings` section of the RTD parcels project panel?
image

**Comment by erikvansebille** [comment]
> Just confirming @erikvansebille , is this option checked in the `Advanced Settings` section of the RTD parcels project panel?

Yep, that checkbox is on (see below), but the default branch is not set. Should I do that?

Screenshot 2023-03-27 at 19 12 38

**Comment by VeckoTheGecko** [comment]
Following on from [this (10 year old) conversation](https://github.com/readthedocs/readthedocs.org/issues/239), it doesn't look like builds are triggered from PRs originating from branches on forks. No matter though. My RTD build for this branch is available [here](https://github.com/readthedocs/readthedocs.org/issues/239). We'll review properly when merging `new_docs` -> `master`.


> the default branch is not set. Should I do that?
No need 😄

**Comment by erikvansebille** [comment]
Thanks for digging into this

> My RTD build for this branch is available [here](https://github.com/readthedocs/readthedocs.org/issues/239). We'll review properly when merging `new_docs` -> `master`.

That link seems to point to the RTD Issue? Can you update to the correct link?

**Comment by VeckoTheGecko** [comment]
Opps, yes, just updated

**Comment by VeckoTheGecko** [comment]
Used the regex `\s:$` to search across the codebase for parameters with missing types. Added many types, will leave the rest to be gradually done over a longer timeframe.

PR ready for feedback 😄

File: parcels/fieldset.py [hunk 1]
```diff
@@ -426,62 +482,76 @@ def from_nemo(cls, filenames, variables, dimensions, indices=None, mesh='spheric
...
+            Watch out: NEMO is discretised on a C-grid:
+            U and V velocities are not located on the same nodes (see https://www.nemo-ocean.eu/doc/node19.html ).
+
+            +-----------------------------+-----------------------------+-----------------------------+
+            |                             |         V[k,j+1,i+1]        |                             |
+            +-----------------------------+-----------------------------+-----------------------------+
+            |U[k,j+1,i]                   |W[k:k+2,j+1,i+1],T[k,j+1,i+1]|U[k,j+1,i+1]                 |
+            +-----------------------------+-----------------------------+-----------------------------+
+            |                             |         V[k,j,i+1]          +                             |
+            +-----------------------------+-----------------------------+-----------------------------+
```
Comment (erikvansebille) [inline parcels/fieldset.py:519]: These tables don't show as nice as they do in the old version? Compare https://veckoparcels.readthedocs.io/en/docstrings/#parcels.fieldset.FieldSet.from_c_grid_dataset to https://parcels.readthedocs.io/en/latest/#parcels.fieldset.FieldSet.from_c_grid_dataset

**Comment by VeckoTheGecko** [comment]
> The title in your https://veckoparcels.readthedocs.io/en/docstrings/ version mentions `v2.3.0`; will that be the latest after we merge?

As of 398ba3b the version number is determined automatically using the `parcels.__version__` attribute. `new_docs` (the target of this PR) is at `v2.3.0`. When we re-sync and merge `new_docs` into `master` it will be at the intended version.

> The copyright statement on the page bottom could/should be updated to 2023?

Can do. Should we dynamically determine the year for the copyright statement? i.e. update the copyright statement from:

```py
copyright = '2022, The OceanParcels Team'
```
to
```py
copyright = f"{datetime.datetime.now().year}, The OceanParcels Team"
```

Happy to fix this in a separate PR.

> Should we add these commits (especially the really large ones) to the `.git-blame-ignore-revs` file too?

I don't think so. The purpose of `.git-blame-ignore-revs` is to ignore (large) *superfluous* changes that in the end have no effect on the function of the code (eg. introducing black, introducing isort, whitespace. These are mainly created by automated formatting tools) and would otherwise soil the blame. These commits mark a shift in how the codebase works with docstrings, and introduce additional edits to the docstrings, which are worth keeping in blame.

**Comment by erikvansebille** [comment]
OK is fine, let's then just go for it. The integration tests have skipped (not sure why), but I guess we can merge without these in this case, because only docstrings are affected in this PR?

**Comment by VeckoTheGecko** [comment]
> The integration tests have skipped (not sure why)

#1335 rearing its head again 😕

And yes, only docstrings/doc related changes.

**Pull request merged**